### PR TITLE
add support for creating variation fonts

### PR DESCRIPTION
**Pull request by lsalzman**
This is necessary to fix downstream Gecko bug https://bugzilla.mozilla.org/show_bug.cgi?id=1441323

### CODE REVIEW

File: src/font_face.rs [hunk 1]
```diff
@@ -270,6 +272,43 @@ impl FontFace {
             (*self.native.get()).GetIndex()
         }
     }
+
+    pub fn has_variations(&self) -> bool {
+        unsafe {
+            let face5: Option<ComPtr<IDWriteFontFace5>> = (*self.native.get()).query_interface(&IDWriteFontFace5::uuidof());
```
Comment (kvark) [inline src/font_face.rs]: nit: can just match on it

File: src/font_face.rs [hunk 2]
```diff
@@ -270,6 +272,43 @@ impl FontFace {
...
+            }
+            false
+        }
+    }
+
+    pub fn create_font_face_with_variations(
+        &self,
+        simulations: DWRITE_FONT_SIMULATIONS,
+        axis_values: &[DWRITE_FONT_AXIS_VALUE],
+    ) -> Option<FontFace> {
```
Comment (kvark) [inline src/font_face.rs:302]: can we have some sort of an error code here (and return `Result`)?

File: src/font_face.rs [hunk 3]
```diff
@@ -270,6 +272,43 @@ impl FontFace {
...
+        unsafe {
+            let face5: Option<ComPtr<IDWriteFontFace5>> = (*self.native.get()).query_interface(&IDWriteFontFace5::uuidof());
```
Comment (kvark) [inline src/font_face.rs]: is this how extension methods are discovered in dwrote? I wonder if the query_interface would be better to call at the construction site and just cache this `ComPtr<IDWriteFontFace5>` somewhere.

**Comment by lsalzman** [comment]
I added support for caching the FontFace5 on demand - I don't want to impose the cost of querying it on all fonts, just the ones that need it for variations.

As for errors, the errors are mostly dealt with ahead of time in Gecko, and we just need a pass or fail here with Option being what is most consistent with dwrote. This is a niche case where I do not believe having to construct a new error type just for it would add any value, since it would require decoding HRESULTs and other nastiness. And I don't know there's anything we can do with the HRESULT either.

**Comment by kvark** [comment]
>I don't want to impose the cost of querying it on all fonts, just the ones that need it for variations.

Is it really going to show up in any profiles though? I imagine we do hundreds of COM calls on a font, so querying the interface at load time shouldn't be a concern.

**Comment by lsalzman** [comment]
It mirrors the way we deal with it in Gecko. There's no need to incur the
expense when for 99% of fonts it is not needed at all.

**Review by kvark — APPROVED** [review]
Would be great if somebody from regular reviewers would also take a look before merging.

**Comment by kvark** [comment]
hmm, doesn't look like there are permanent reviewers
@bors-servo r+

**Pull request merged**